### PR TITLE
Adds `attempt` decorator

### LINKED ISSUE #264: Variant of safe that preserves the argument that lead to an exception
I propose a function/decorator (maybe `attempt` would be a good name?) similar to `safe` with the twist of returning `Failure(bad_input)` from `safe(f)(bad_input)` instead of `Failure(SomeException)`. An example where this would be useful:

```python
(
    attempt(json.loads)('<hello world>')
    .alt(lambda s: ValueError(f'"{s}" is not a valid json')) 
)
# Failure(ValueError('"<hello world>" is not a valid json'))
```

### PR DESCRIPTION
**Pull request by thepabloaguilar**
# Adds `attempt` decorator

## Checklist

- [X] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [X] I have created at least one test case for the changes I have made
- [X] I have updated the documentation for the changes I have made
- [X] I have added my changes to the `CHANGELOG.md`

## Related issues

Closes #264

### CODE REVIEW

File: docs/pages/result.rst [hunk 1]
```diff
@@ -116,6 +116,28 @@ If you want to `safe` handle only a set of exceptions:
     ...
   ValueError: Too big
 
+attempt
+~~~~~~~
+
+Similar to :func:`safe <returns.result.safe>` function but instead
+of wrapping the exception error in a Failure container it'll wrap the
```
Comment (sobolevn) [inline docs/pages/result.rst:123]:
```suggestion
of wrapping the exception error in a :func:`~Failure` container it'll wrap the
```

File: typesafety/test_result/test_attempt.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+- case: safe_decorator_one_param
+  disable_cache: false
+  main: |
+    from returns.result import attempt
+
+    @attempt
+    def test(param: str) -> int:
+        return 1
+
+    reveal_type(test)  # N: Revealed type is "def (builtins.str*) -> returns.result.Result[builtins.int*, builtins.str*]"
```
Comment (sobolevn) [inline typesafety/test_result/test_attempt.yml:22]: Let's also add this example:

```python
def other(param: int) -> str:
    return 'a'

reveal_type(attempt(other))
```

**Review by sobolevn — COMMENTED** [review]
Thanks you! Only minor suggestions.

I've also removed a test file, because doctests are good enough in this case.

File: docs/pages/result.rst (hunk 1 above)
Comment (thepabloaguilar) [reply docs/pages/result.rst]: I've added a class reference since the exposed `Failure` is a class:
```text
:class:`Failure <returns.result.Failure>`
```

**Review by sobolevn — APPROVED** [review]
Great work! Thank you!

**Pull request merged**